### DISCUSSION TITLE
Req.exe is crashing with error "OPENSSL_Uplink(00007FFE08E3F5F8,08): no OPENSSL_Applink" on Windows Server 2022 VM

### DISCUSSION THREAD

**Discussion by gsharma-ad** [openssl/openssl#24244] [Q&A]
Req.exe is crashing with error "_OPENSSL_Uplink(00007FFE08E3F5F8,08): no OPENSSL_Applink_" on Windows Server 2022 VM as below-:

    C:\utilities>req.exe -out clientreq -newkey rsa:2048 -keyout mykey -verbose
    Using configuration from C:\Program Files\Common Files\SSL/openssl.cnf
    Generating rsa key with 2048 bits
    ......
    Writing private key to 'mykey'
    OPENSSL_Uplink(00007FFE0914F5F8,08): no OPENSSL_Applink
    C:\utilities>
Req.exe is compiled with MSVC2022.

When I made below changes in the apps.c at line 3071 (_in function bio_open_owner_) -:

    //b = BIO_new_fp(fp, bflags);   // commented this line
    b = BIO_new_file(filename, "w"); // Added this line
Crash issue is resolved. 

Please suggest solution for this behavior. If above change is ok, then I can raise a pull request as well for the same.

**Comment by mattcaswell** [comment]
How did you create req.exe? That isn't a standard OpenSSL executable - that is called openssl.exe.

**Comment by gsharma-ad** [comment]
In our product, a Makefile is written to copy req.c and other required files at a separate folder and compile the same to generate the req.exe.

**Comment by mattcaswell** [comment]
You should compile and link uplink.c into your req.exe:

https://github.com/openssl/openssl/blob/master/ms/uplink.c


See:
https://www.openssl.org/docs/man3.3/man3/OPENSSL_Applink.html

**Comment by gsharma-ad** [comment]
I have added uplink.c and its dependent files to created req.exe, but now I am getting below error-:

    uplink.obj : error LNK2019: unresolved external symbol OPENSSL_showfatal referenced in function unimplemented
    test_Windows_NT\reqtest.exe : fatal error LNK1120: 1 unresolved externals
    make[6]: *** [../../../../../make/Makefile.test:68: test_Windows_NT/reqtest.exe] Error 2
_OPENSSL_showfatal_ symobl not present in libcrypo.lib and libssl.lib.

**Comment by tom-cosgrove-arm** [comment]
> In our product, a Makefile is written to copy req.c and other required files

> OPENSSL_showfatal symobl not present in libcrypo.lib and libssl.lib

`OPENSSL_showfatal` comes from `crypto/cryptlib.c`

NOTE: You are unlikely to get much more help trying to debug a completely non-standard build of OpenSSL

**Reply by gsharma-ad** [reply]
I am able to created req.exe after adding libcrypt.c along with other files. And provided option -DOPENSSL_USE_APPLINK during compilation. Req.exe still crashing with - OPENSSL_Uplink(00007FFE0914F5F8,08): no OPENSSL_Applink.
Behavior is same without compilation option -DOPENSSL_USE_APPLINK.

**Reply by gsharma-ad** [reply]
On Windows, I was getting the same error with OpenSSL libraries with other application which is using libcrypto.dll and libssl.dll. There I fixed with below same changes-:
```
//b = BIO_new_fp(fp, bflags);   // commented this line
b = BIO_new_file(filename, "w"); // Added this line
